packages/transformers/js/src/visitors/dependencies.js

```diff
@@ -291,11 +292,19 @@ function addDependency(
   });
 }
 
-function addURLDependency(asset, node, opts = {}) {
-  node.value = asset.addURLDependency(node.value, {
+function addWorkerDependency(asset, node, opts = {}) {
+  let url = node.value;
+  asset.addURLDependency(url, {
     loc: node.loc && createDependencyLocation(node.loc.start, node.value, 0, 1),
     ...opts,
   });
+  morph(
+    node,
+    types.callExpression(types.identifier('require'), [
+      types.stringLiteral(url),
+    ]),
+  );
+  node.skipDependency = true;
```

Comment:
rewriting this to a `require` causes another dependency to be added during this same traversal. Any alternative @devongovett @mischnic?

Reply:
Yes, turning the `CallExpression` visitor into `CallExpression: { exit(node) {....`.
That way, you are replacing the children after the children were already processed.

Reply:
thanks 😄

Reply:
Actually, that doesn't seem to work. I think since this transformation creates a new CallExpression that new node then gets traversed (it's not this node that causes the new dependency).

Reply:
Looks like that won't work, unfortunately. It's not this node (the original CallExpression) that creates the new dependency, it's the argument StringLiteral node that gets transformed into another CallExpression that does. It's caused by the entry visitor being called again (this time on the morphed node) during this traversal.

Reply:
I tested it before suggesting and it worked?
```js
const parser = require("@babel/parser");
const generate = require("@babel/generator").default;
const types = require("@babel/types");
const { ancestor } = require("babylon-walk");

function morph(object, newProperties) {
	for (let key in object) {
		delete object[key];
	}
	for (let key in newProperties) {
		object[key] = newProperties[key];
	}
}

let ast = parser.parse(`require("test!");`);
ancestor(ast, {
	CallExpression: {
		// crashes: enter(node) {
		exit(node) {
			let { callee, arguments: args } = node;

			morph(
				args[0],
				types.callExpression(types.identifier("require"), [
					types.stringLiteral("URL")
				])
			);
		}
	}
});
console.log(generate(ast).code);
```

Reply:
Ah, good catch. It looks like the `NewExpression` visitor below also takes the path which morphs the node. I've updated that to use an exit visitor.